### PR TITLE
Don't update condition if status stays False/Unknown for custom plugin

### LINKED ISSUE #202: Reason change once and it stay in wrong state with custom plugins
Hello,

I found something again I think but maybe I follwed wrong way. 
I'm using a custom config based on the ntp example: 

```
{
    "plugin": "custom",
    "pluginConfig": {
        "invoke_interval": "30s",
        "timeout": "5s",
        "max_output_length": 80,
        "concurrency": 3
    },
    "source": "ntp-custom-plugin-monitor",
    "conditions": [
        {
            "type": "CustomProblem",
            "reason": "CustomIsUp",
            "message": "Status of the custom service"
        }
    ],
    "rules": [
        {
            "type": "permanent",
            "condition": "CustomProblem",
            "reason": "CustomIsDown",
            "path": "/usr/bin/custom.sh",
            "timeout": "3s"
        }
    ]
}

```
The /usr/bin/custom.sh script is very simple: exit with 0 or 1. 

So when the node problem detector start it set the condition:

```
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  CustomProblem    False   Wed, 05 Sep 2018 14:40:26 +0200   Wed, 05 Sep 2018 14:40:25 +0200   CustomIsUp                   Status of the custom service
  OutOfDisk        False   Wed, 05 Sep 2018 14:40:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Wed, 05 Sep 2018 14:40:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 05 Sep 2018 14:40:23 +0200   Wed, 05 Sep 2018 13:17:06 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 05 Sep 2018 14:40:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 05 Sep 2018 14:40:23 +0200   Thu, 30 Aug 2018 17:35:04 +0200   KubeletReady                 kubelet is posting ready status

```
After it run the script (what returned with 0 in this case) the Status stay false but the Reason field changed to what I set in the rule section:

```
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  CustomProblem    False   Wed, 05 Sep 2018 14:41:56 +0200   Wed, 05 Sep 2018 14:40:55 +0200   CustomIsDown                 Status of the custom service
  OutOfDisk        False   Wed, 05 Sep 2018 14:42:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Wed, 05 Sep 2018 14:42:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 05 Sep 2018 14:42:23 +0200   Wed, 05 Sep 2018 13:17:06 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 05 Sep 2018 14:42:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 05 Sep 2018 14:42:23 +0200   Thu, 30 Aug 2018 17:35:04 +0200   KubeletReady                 kubelet is posting ready status
```

So ok, in the next run the script exited with 1. The Status is True, and the Reason still same (this is what I set under the rule):

```
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  CustomProblem    True    Wed, 05 Sep 2018 14:43:56 +0200   Wed, 05 Sep 2018 14:43:55 +0200   CustomIsDown                 Status of the custom service
  OutOfDisk        False   Wed, 05 Sep 2018 14:43:53 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Wed, 05 Sep 2018 14:43:53 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 05 Sep 2018 14:43:53 +0200   Wed, 05 Sep 2018 13:17:06 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 05 Sep 2018 14:43:53 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 05 Sep 2018 14:43:53 +0200   Thu, 30 Aug 2018 17:35:04 +0200   KubeletReady                 kubelet is posting ready status
```

In the next round the script returned with 0 again and Status changed back to false but the Reason didn't change: 

```
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  CustomProblem    False   Wed, 05 Sep 2018 14:44:26 +0200   Wed, 05 Sep 2018 14:44:25 +0200   CustomIsDown                 Status of the custom service
  OutOfDisk        False   Wed, 05 Sep 2018 14:44:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Wed, 05 Sep 2018 14:44:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Wed, 05 Sep 2018 14:44:23 +0200   Wed, 05 Sep 2018 13:17:06 +0200   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Wed, 05 Sep 2018 14:44:23 +0200   Thu, 30 Aug 2018 17:16:47 +0200   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Wed, 05 Sep 2018 14:44:23 +0200   Thu, 30 Aug 2018 17:35:04 +0200   KubeletReady                 kubelet is posting ready status
```

As I see you overwrite the condition's rule and maybe the original condition lost and the node problem detector never can't set it again. [https://github.com/kubernetes/node-problem-detector/blob/master/pkg/custompluginmonitor/custom_plugin_monitor.go#L140](https://github.com/kubernetes/node-problem-detector/blob/master/pkg/custompluginmonitor/custom_plugin_monitor.go#L140)
But maybe I missed something.

Thank you!

### LINKED ISSUE #306: LastHeartbeatTime is not update when plugin does not have a condition
We have a plugin that runs every 5 minutes. Previously it had a `conditions`

```json
  "conditions": [
    {
      "type": "NetworkConnectivityProblem",
      "reason": "NetworkConnectivityNormal",
      "message": "Network connectivity is normal"
    }
  ],
```

which we removed because it causes:
 -  `LastTransitionTime` to be updated on each deploy https://github.com/kubernetes/node-problem-detector/issues/304
 - `LastHeartbeatTime` to be updated every minute even if it did not run

... but now it does not update `LastHeartbeatTime` at all, even though we see in the logs that it runs (succeeds) every 5 minutes. So we have to chose between 2 broken states ... please add a fix or allow some config like `heartBeatMakesSense: true` that updates the heartbeat only when the plugin runs 🤷‍♂ 

FYI rest of the plugin config
```json
{
  "plugin": "custom",
  "pluginConfig": {
    "invoke_interval": "5m",
    "timeout": "60s",
    "max_output_length": 80,
    "concurrency": 1
  },
  "source": "node-network-connectivity",
  "rules": [
    {
      "type": "permanent",
      "condition": "NetworkConnectivityProblem",
      "reason": "NetworkConnectivityFail",
      "path": "./config/plugin/network_connectivity.rb",
      "timeout": "60s"
    }
  ]
}
```
/cc @eatwithforks @xueweiz @wangzhen127

### PR DESCRIPTION
**Pull request by wangzhen127**
Fixes #202

When condition status stays True, we can update condition if reason or message changes. But when condition status stays False/Unknown, we should just ignore and do not update condition.

Here is the example:
1. NPD starts up, custom condition is set to False and reason is set to default reason.
2. When custom plugin runs, status stays False, but `result.Rule.Reason` is always different from the existing reason, which is the default reason set in step 1. In this case, we should just ignore and do not update condition.

Also added some validation to check custom plugin config.
Fixes #306

### CODE REVIEW

File: pkg/custompluginmonitor/custom_plugin_monitor.go [hunk 1]
```diff
@@ -160,76 +160,68 @@ func (c *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Stat
...
-						result.Rule.Reason,
-						timestamp,
-					))
-
-					condition.Status = status
-					condition.Reason = result.Rule.Reason
-				} else if condition.Status == status &&
+					// Scenario 3: Condition status changes from False to Unknown or vice versa
+					newReason = defaultConditionReason
+					newMessage = defaultConditionMessage
```
Comment (wangzhen127) [inline pkg/custompluginmonitor/custom_plugin_monitor.go]: @Random-Liu @andyxning 
This is different from previous implementation. I think it makes more sense to set to default reason and message here. And unknow status is somewhat undefined. So I don't think there will be any compatibility issue.

WDYT?

File: pkg/custompluginmonitor/types/config.go [hunk 2]
```diff
@@ -141,5 +141,22 @@ func (cpc CustomPluginConfig) Validate() error {
...
+		conditionType := rule.Condition
+		defaultConditionExists := false
+		for _, cond := range cpc.DefaultConditions {
+			if conditionType == cond.Type {
+				defaultConditionExists = true
+				break
+			}
+		}
+		if !defaultConditionExists {
+			return fmt.Errorf("Permanent problem %s does not have preset default condition.", conditionType)
```
Comment (grosser) [inline pkg/custompluginmonitor/types/config.go]: awesome!

File: pkg/custompluginmonitor/types/config.go [hunk 3]
```diff
@@ -141,5 +141,22 @@ func (cpc CustomPluginConfig) Validate() error {
 		}
 	}
 
+	for _, rule := range cpc.Rules {
```
Comment (grosser) [inline pkg/custompluginmonitor/types/config.go]: maybe a `validateDefaultConditionExists` method ?

File: pkg/custompluginmonitor/custom_plugin_monitor.go [hunk 4]
```diff
@@ -160,73 +160,62 @@ func (c *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Stat
...
-					var defaultConditionReason string
-					var defaultConditionMessage string
-					for j := range c.config.DefaultConditions {
-						defaultCondition := &c.config.DefaultConditions[j]
-						if defaultCondition.Type == result.Rule.Condition {
-							defaultConditionReason = defaultCondition.Reason
-							defaultConditionMessage = defaultCondition.Message
-							break
-						}
+				var defaultConditionReason string
```
Comment (grosser) [inline pkg/custompluginmonitor/custom_plugin_monitor.go]: some comment about what is going on in that block ... or even extract a method
`// find default condition`

File: pkg/custompluginmonitor/custom_plugin_monitor.go [hunk 5]
```diff
@@ -160,73 +160,62 @@ func (c *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Stat
...
-
-					condition.Status = status
-					condition.Message = defaultConditionMessage
-					condition.Reason = defaultConditionReason
+				needToUpdateCondition := true
+				var newReason string
+				var newMessage string
+				status := toConditionStatus(result.ExitStatus)
+				if condition.Status == types.True && status != types.True {
+					// Scenario 1: Condition status changes from True to False/Unknown
```
Comment (grosser) [inline pkg/custompluginmonitor/custom_plugin_monitor.go]: some more enduser centered docs would be nice like
```
// Scenario 1: status recovered (True -> False/Unknown), switch it back to the default
```

**Comment by grosser** [comment]
logic looks correct to me, I have the feeling there is a more logical ordering for the if/else, like first going over all the `status == True / != True`, but should work either way 👍

**Comment by grosser** [comment]
might be better to turn the new validation into it's own PR so that it can be merged more quickly (also needs a test ?)

File: pkg/custompluginmonitor/custom_plugin_monitor.go (hunk 4 above)
Comment (wangzhen127) [reply pkg/custompluginmonitor/custom_plugin_monitor.go]: Added some comment

File: pkg/custompluginmonitor/types/config.go (hunk 3 above)
Comment (wangzhen127) [reply pkg/custompluginmonitor/types/config.go]: I think either way is fine. I will just keep it here as I feel this is still simple.

File: pkg/custompluginmonitor/custom_plugin_monitor.go (hunk 5 above)
Comment (wangzhen127) [reply pkg/custompluginmonitor/custom_plugin_monitor.go]: This logic is very complex. We have been thinking about improving it. This CL is more like a band-aid. So I will just keep the doc and probably improve the whole logic later.

**Comment by grosser** [comment]
this bot is annoying ... there is commit status for that :/

**Pull request merged**